### PR TITLE
Breakup hooks into their own files

### LINKED ISSUE #209: Refactor hooks into their own files
app.js is 500 lines of badly organized javascript.

Let's instead it lots of small files :)

### PR DESCRIPTION
**Pull request by teesloane**
This should not be merged  until #208 merges, as this branches off of `mv/simplify_map`.

Some hooks have been renamed to be more specific. Example:  the "Chart" hook is now StatsDashboardHook.

There are a few things in app.js that don't see to be used - I've tagged them with `REVIEW` comments

closes #209

### CODE REVIEW

**Comment by mveytsman** [comment]
I think this is ready for review? Do you mind rebasing it against main again so we have a clean PR?

File: assets/js/app.js [hunk 1]
```diff
@@ -578,6 +98,7 @@ liveSocket.connect()
 // >> liveSocket.enableLatencySim(1000)
 window.liveSocket = liveSocket
 
+// REVIEW: I can't find hidePII or showPII used anywhere. Can we remove?
```
Comment (mveytsman) [inline assets/js/app.js:101]: I use it from the browser JS console during demos to blur out PII 😁 

I know its weird but I couldn't think of a better way

File: assets/js/app.js [hunk 2]
```diff
@@ -15,234 +15,12 @@ import L from "leaflet"
...
-      if (this.layers[e.detail.id]) {
-        this.layers[e.detail.id].removeFrom(this.map)
-        delete this.layers[e.detail.id]
-      }
-    })
-  },
-};
+let Hooks = hooks
 
+//REVIEW: LeafletLayer is not being used anywhere? Can I remove?
```
Comment (mveytsman) [inline assets/js/app.js:23]: Yup this should be cut

File: assets/js/hooks/index.js [hunk 3]
```diff
@@ -0,0 +1,31 @@
...
+  ConversationList,
+  MessageList,
+  RiderSelectionList,
+  Alpine,
+  Autocomplete,
+  EmojiButtonHook,
+  StatsDashboardChart,
+  CheckboxAll,
+  FrameHook,
+  // REVIEW: this hook is used on TagsComponent module, which isn't used anywhere. Remove it?
```
Comment (mveytsman) [inline assets/js/hooks/index.js:28]: 😬 https://github.com/bikebrigade/dispatch/commit/9781f9d723507275e076994c3f14359127f97d3f

File: assets/js/app.js (hunk 1 above)
Comment (teesloane) [reply assets/js/app.js:101]: What's PII?

**Comment by mveytsman** [comment]
Personally identifying information

If you search the code base for the class this thing blurs it’s put on like
phone numbers and addresses so they get blurred out for eg a screenshot we
want to show someone

**Comment by teesloane** [comment]
> I think this is ready for review? Do you mind rebasing it against main again so we have a clean PR?

merged in the updates (sorry, no rebase this time) and should be ready for re-review.

File: assets/js/app.js [hunk 4]
```diff
@@ -476,6 +59,7 @@ liveSocket.connect()
 // >> liveSocket.enableLatencySim(1000)
 window.liveSocket = liveSocket
 
+// REVIEW: I can't find hidePII or showPII used anywhere. Can we remove?
```
Comment (mveytsman) [inline assets/js/app.js:62]:
```suggestion
```

File: assets/js/hooks/alpine.js [hunk 5]
```diff
@@ -0,0 +1,12 @@
+export default {
```
Comment (mveytsman) [inline assets/js/hooks/alpine.js:1]:
```suggestion

const Alpine = {
```

File: assets/js/hooks/alpine.js [hunk 6]
```diff
@@ -0,0 +1,12 @@
...
+    this.el.dispatchEvent(new CustomEvent('mounted', {
+      bubbles: true
+    }));
+  },
+  updated() {
+    this.el.dispatchEvent(new CustomEvent('updated', {
+      bubbles: true
+    }));
+  },
+}
```
Comment (mveytsman) [inline assets/js/hooks/alpine.js:12]:
```suggestion
};

export default Alpine;
```

File: assets/js/hooks/autocomplete.js [hunk 7]
```diff
@@ -0,0 +1,18 @@
+import Tribute from "tributejs";
+
+export default {
```
Comment (mveytsman) [inline assets/js/hooks/autocomplete.js:3]:
```suggestion
const Autocomplete = {
```

File: assets/js/hooks/autocomplete.js [hunk 8]
```diff
@@ -0,0 +1,18 @@
...
+        return {
+          key: s,
+          value: `{{${s}}}}`
+        }
+      })
+    });
+
+    tribute.attach(this.el)
+  }
+}
```
Comment (mveytsman) [inline assets/js/hooks/autocomplete.js:18]:
```suggestion
};

export default Autocomplete;
```

File: assets/js/hooks/checkbox_all.js [hunk 9]
```diff
@@ -0,0 +1,17 @@
+export default {
```
Comment (mveytsman) [inline assets/js/hooks/checkbox_all.js:1]:
```suggestion
const CheckboxAll = {
```

File: assets/js/hooks/checkbox_all.js [hunk 10]
```diff
@@ -0,0 +1,17 @@
...
+      this.el.indeterminate = false;
+    } else if (numChecked === numRows) {
+      this.el.checked = true;
+      this.el.indeterminate = false;
+    } else {
+      this.el.checked = false;
+      this.el.indeterminate = true;
+    }
+  }
+}
```
Comment (mveytsman) [inline assets/js/hooks/checkbox_all.js:17]:
```suggestion
};

export default CheckboxAll;
```

File: assets/js/hooks/conversation_list.js [hunk 11]
```diff
@@ -0,0 +1,28 @@
+export default {
```
Comment (mveytsman) [inline assets/js/hooks/conversation_list.js:1]:
```suggestion
const ConversationList = {
```

File: assets/js/hooks/conversation_list.js [hunk 12]
```diff
@@ -0,0 +1,28 @@
...
+    this.handleEvent("new_message", ({
+      riderId
+    }) => {
+      let msg = document.getElementById(`conversation-list-item:${riderId}`);
+      if (msg != undefined) {
+        this.el.prepend(msg)
+      }
+    });
+  }
+}
```
Comment (mveytsman) [inline assets/js/hooks/conversation_list.js:28]:
```suggestion
};

export default ConversationList;
```

File: assets/js/hooks/emoji_button.js [hunk 13]
```diff
@@ -0,0 +1,24 @@
+import { EmojiButton } from '@joeattardi/emoji-button';
+
+export default {
```
Comment (mveytsman) [inline assets/js/hooks/emoji_button.js:3]:
```suggestion
const EmojiButtonHook = {
```

File: assets/js/hooks/emoji_button.js [hunk 14]
```diff
@@ -0,0 +1,24 @@
...
+
+    picker.on("hidden", () => {
+      const end = inputEl.value.length;
+      inputEl.setSelectionRange(end, end);
+      inputEl.focus();
+    });
+    this.el.addEventListener("click", () => picker.togglePicker(this.el));
+  }
+
+}
```
Comment (mveytsman) [inline assets/js/hooks/emoji_button.js:24]:
```suggestion
};

export default EmojiButtonHook;
```

File: assets/js/hooks/frame.js [hunk 15]
```diff
@@ -0,0 +1,14 @@
+export default {
```
Comment (mveytsman) [inline assets/js/hooks/frame.js:1]:
```suggestion
const FrameHook = {
```

File: assets/js/hooks/frame.js [hunk 16]
```diff
@@ -0,0 +1,14 @@
...
+      window.parent.postMessage({
+        height: this.el.scrollHeight
+      }, "*");
+    });
+
+    resizeObserver.observe(this.el);
+    window.parent.postMessage({
+      height: this.el.scrollHeight
+    }, "*");
+  }
```
Comment (mveytsman) [inline assets/js/hooks/frame.js:13]:
```suggestion
  };
  
  export default FrameHook;
```

File: assets/js/hooks/leaflet_map.js [hunk 17]
```diff
@@ -0,0 +1,159 @@
+export default {
```
Comment (mveytsman) [inline assets/js/hooks/leaflet_map.js:1]:
```suggestion
const LeafletMap = {
```

File: assets/js/hooks/leaflet_map.js [hunk 18]
```diff
@@ -0,0 +1,159 @@
...
+        let lng = this.el.dataset.lng || "-79.3832"
+        this.map.setView([lat, lng], this.el.dataset.zoom || 13);
+      }
+
+      this.map.invalidateSize()
+
+    });
+  },
+};
+
```
Comment (mveytsman) [inline assets/js/hooks/leaflet_map.js:159]:
```suggestion

export default LeafletMap;
```

File: assets/js/hooks/message_list.js [hunk 19]
```diff
@@ -0,0 +1,23 @@
...
+          this.el.scrollTop = this.el.scrollHeight - curScrollheight;
+        });
+      }
+    });
+    this.handleEvent("new_message", () => {
+      // scroll to bottom
+      this.el.scrollTop = this.el.scrollHeight;
+    });
+  }
+}
```
Comment (mveytsman) [inline assets/js/hooks/message_list.js:23]:
```suggestion
};

export default MessageList;
```

File: assets/js/hooks/riders_list.js [hunk 20]
```diff
@@ -0,0 +1,13 @@
+export default {
```
Comment (mveytsman) [inline assets/js/hooks/riders_list.js:1]:
```suggestion
const RidersList = {
```

File: assets/js/hooks/riders_list.js [hunk 21]
```diff
@@ -0,0 +1,13 @@
...
+      id
+    }) => {
+      this.el.querySelector(`[id="riders-list:${id}"]`).scrollIntoView({
+        behavior: 'auto',
+        block: 'nearest',
+        inline: 'start'
+      })
+    })
+  }
+}
```
Comment (mveytsman) [inline assets/js/hooks/riders_list.js:13]:
```suggestion
};

export default RidersList;
```

File: assets/js/hooks/stats_dashboard_chart.js [hunk 22]
```diff
@@ -0,0 +1,21 @@
+import Chart from 'chart.js/auto';
+
+export default {
```
Comment (mveytsman) [inline assets/js/hooks/stats_dashboard_chart.js:3]:
```suggestion
const StatsDashboardChart = {
```

File: assets/js/hooks/stats_dashboard_chart.js [hunk 23]
```diff
@@ -0,0 +1,21 @@
...
+        }
+      }
+    });
+
+    this.handleEvent("update_chart", (data) => {
+      this.chart.data = data;
+      this.chart.update();
+    })
+  }
+}
```
Comment (mveytsman) [inline assets/js/hooks/stats_dashboard_chart.js:21]:
```suggestion
};

export default StatsDashboardChart;
```

File: assets/js/hooks/tags_component.js [hunk 24]
```diff
@@ -0,0 +1,13 @@
+export default {
```
Comment (mveytsman) [inline assets/js/hooks/tags_component.js:1]:
```suggestion
const TagsComponentHook = {
```

File: assets/js/hooks/tags_component.js [hunk 25]
```diff
@@ -0,0 +1,13 @@
...
+      if (e.key == 'Enter') {
+        e.preventDefault();
+        this.pushEventTo(this.el, 'select', {
+          name: this.el.value
+        })
+        this.el.value = ""
+      }
+    });
+  }
+}
```
Comment (mveytsman) [inline assets/js/hooks/tags_component.js:13]:
```suggestion
};

export default TagsComponentHook;
```

File: assets/js/hooks/tasks_list.js [hunk 26]
```diff
@@ -0,0 +1,13 @@
+export default {
```
Comment (mveytsman) [inline assets/js/hooks/tasks_list.js:1]:
```suggestion
const TasksList = {
```

File: assets/js/hooks/tasks_list.js [hunk 27]
```diff
@@ -0,0 +1,13 @@
...
+      id
+    }) => {
+      this.el.querySelector(`[id="tasks-list:${id}"]`).scrollIntoView({
+        behavior: 'auto',
+        block: 'nearest',
+        inline: 'start'
+      })
+    })
+  }
+}
```
Comment (mveytsman) [inline assets/js/hooks/tasks_list.js:13]:
```suggestion
};

export default TasksList;
```

File: assets/js/app.js [hunk 28]
```diff
@@ -2,396 +2,18 @@
...
-import {
-  LiveSocket
-} from "phoenix_live_view"
-
-
+import { LiveSocket } from "phoenix_live_view"
 import L from "leaflet"
 import "leaflet-makimarkers"
 import "@ryangjchandler/alpine-clipboard"
 import "alpinejs"
```
Comment (mveytsman) [inline assets/js/app.js:11]: Should these imports be in the relevant files?

**Review by mveytsman — APPROVED** [review]
Looks great!

I added some suggested edits to explicitly name each hook in it's file -- I think this will make finding the right file to edit later on easier.

I think it would be nice to make the naming consistent, and rename `EmojiButtonHook` to `EmojiButton`, `FrameHook` to `Frame` and `TagsComponentHook` to `TagsComponent` but we can do that in another PR too...

Otherwise 🚀

File: assets/js/app.js (hunk 28 above)
Comment (teesloane) [reply assets/js/app.js:11]: Yes, I should have removed these from app.js.

Comment (teesloane) [reply assets/js/app.js]: Looks like alpine and alpine clipboard should be left, as it seems to register it globally when imported like this and not like [described here](https://github.com/ryangjchandler/alpine-clipboard#npm).

File: assets/js/hooks/index.js (hunk 3 above)
Comment (Blineberger) [reply assets/js/hooks/index.js]: - [ ] ?

Comment (Blineberger) [reply assets/js/hooks/index.js]: > 😬 https://github.com/bikebrigade/dispatch/commit/9781f9d723507275e076994c3f14359127f97d3f

**Pull request merged**